### PR TITLE
Support for time specification for Kerberos ticket expiration (ticket…

### PR DESCRIPTION
**Pull request by kirtiv1**
…_lifetime and max_life)

### CODE REVIEW

**Comment by santhoshkumarml** [comment]
Thanks a lot for these changes @kirtiv1, the changes look good

**Review by dimaspivak — COMMENTED** [review]
Perhaps this should be merged into the master branch and cherry-picked into streamsets?
--> Right you are Dima. Will do.

File: cm.py [hunk 1]
```diff
@@ -315,6 +315,14 @@ def deploy_cluster_kerberos_client_config(self, cluster_name, host_ids=None):
         }
         return self.api_client.deploy_cluster_kerberos_client_config(cluster_name, host_ref_list)
 
+    def get_cm_kerberos_principals(self):
+        """Get list of Kerberos principals needed by the services being managed by Cloudera Manager.
```
Comment (dimaspivak) [inline cm.py]: A slightly redundant docstring :) Might wanna just have the returns be `A list of strings` or whatever.

File: start.py [hunk 2]
```diff
@@ -345,27 +345,31 @@ def cm_server_not_dead(primary_node):
...
 
     logger.debug('Updating krb5.conf ...')
     krb5_conf = cluster.kdc_node.get_file(KDC_KRB5_CONF_FILENAME)
-    kdc_node.put_file(KDC_KRB5_CONF_FILENAME,
-                      re.sub(r'EXAMPLE.COM', realm,
-                             re.sub(r'example.com', cluster.network,
-                                    re.sub(r'kerberos.example.com',
-                                           kdc_node.fqdn,
-                                           krb5_conf))))
+    ticket_lifetime_repl = kerberos_ticket_lifetime if kerberos_ticket_lifetime else '\g<1>'
```
Comment (dimaspivak) [inline start.py]: Can the `\g<1>` be made a const at the top of the file? (And ideally a comment added for it)

Comment (dimaspivak) [inline start.py]: What is `_repl`? Can we name this variable in a human-readable way (here and below)?

File: start.py [hunk 3]
```diff
@@ -398,7 +407,7 @@ def _configure_kdc(cluster, kerberos_principals, quiet):
     if kerberos_principals:
         kdc_commands.append('chmod 644 {}'.format(KDC_KEYTAB_FILENAME))
 
-    kdc_node.execute('; '.join(kdc_commands),
+    kdc_node.execute('&& '.join(kdc_commands),
```
Comment (dimaspivak) [inline start.py:411]: Awesome change. :D

File: start.py [hunk 4]
```diff
@@ -490,10 +506,13 @@ def failure(timeout):
                        timeout=180,
                        success=success, failure=failure)
 
+    if kerberos_ticket_lifetime:
+        cm_kerberos_principals = deployment.get_cm_kerberos_principals()
+        logger.info('cm_kerberos_principals = %s', cm_kerberos_principals)
```
Comment (dimaspivak) [inline start.py]: This should be `debug` level

File: start.py [hunk 5]
```diff
@@ -490,10 +506,13 @@ def failure(timeout):
...
+        _apply_kerberos_ticket_expiration_for_cm_principals(cluster, cm_kerberos_principals, kerberos_ticket_lifetime)
     for service in deployment.get_cluster_services(cluster_name=DEFAULT_CLUSTER_NAME):
         if service['type'] == 'HUE':
             _apply_kerberos_fix_for_hue(cluster)
-            break
```
Comment (dimaspivak) [inline start.py]: Why was this removed as part of the change?

File: start.py [hunk 6]
```diff
@@ -513,7 +532,22 @@ def _apply_kerberos_fix_for_hue(cluster):
         'service krb5kdc restart',
         'service kadmin restart'
     ]
-    kdc_node.execute('; '.join(kdc_hue_commands))
+    kdc_node.execute('&& '.join(kdc_hue_commands))
+
+
+def _apply_kerberos_ticket_expiration_for_cm_principals(cluster, cm_kerberos_principals, kerberos_ticket_lifetime):
+    """ Apply kerberos ticket expiration for Kerberos principals needed by the services being managed by
```
Comment (dimaspivak) [inline start.py]: No space after opening triple quotes and be consistent with casing of "Kerberos."

File: topology.yaml [hunk 7]
```diff
@@ -65,6 +65,10 @@ start args:
     --kerberos-principals:
         help: If specified, a comma-separated list of Kerberos user principals to create in KDC
         metavar: princ1,princ2,...
+    --kerberos-ticket-lifetime:
+        default: 86400
+        help: If specified, the maximum time period in seconds for which a ticket may be valid in the realm
+        metavar: 600
```
Comment (dimaspivak) [inline topology.yaml]: The `metavar` should probably be something like `seconds` as opposed to a number.

**Review by dimaspivak — CHANGES_REQUESTED** [review]

File: cm.py (hunk 1 above)
Comment (kirtiv1) [reply cm.py]: Done.

File: start.py (hunk 2 above)
Comment (kirtiv1) [reply start.py]: Done.

File: start.py (hunk 4 above)
Comment (kirtiv1) [reply start.py]: Done.

File: start.py (hunk 6 above)
Comment (kirtiv1) [reply start.py]: Done.

File: topology.yaml (hunk 7 above)
Comment (kirtiv1) [reply topology.yaml]: Done.

File: start.py (hunk 2 above)
Comment (kirtiv1) [reply start.py]: Added comment. Thanks !!

File: start.py (hunk 5 above)
Comment (kirtiv1) [reply start.py]: Done.

**Comment by kirtiv1** [comment]
Closing this request as another one was created against master branch which is the correct way to go for this functionality. https://github.com/clusterdock/topology_cdh/pull/25
